### PR TITLE
Fix `MaskedTensor` to device ignored mask

### LINKED ISSUE #147140: Mask in MaskedTensor does not change device
### 🐛 Describe the bug

When you create a MaskedTensor and change it to cuda, the data is the only one that change to cuda. When we use a reduction function on cuda MaskedTensor (sum, to_tensor, etc), it will always fail since the mask in on another device.

```
import torch
from torch.masked import as_masked_tensor
data = torch.tensor([1,2,3])
mask = torch.tensor([True,False,True])
mt = as_masked_tensor(data, mask).to('cuda')
mt.get_data().device, mt.get_mask().device
```
```
(device(type='cuda', index=0), device(type='cpu'))
```

### Versions

Run on torch 2.6, python 3.11

### PR DESCRIPTION
**Pull request by zeshengzong**
Fixes #147140

## Changes

- Add `to` implementation in `MaskedTensor` to support move `mask` to target device

## Test Result

```python
In [1]: import torch
   ...: from torch.masked import as_masked_tensor
   ...: data = torch.tensor([1,2,3])
   ...: mask = torch.tensor([True,False,True])
   ...: mt = as_masked_tensor(data, mask).to('cuda')
   ...: mt.get_data().device, mt.get_mask().device
/home/zong/code/pytorch/torch/masked/maskedtensor/core.py:247: UserWarning: The PyTorch API of MaskedTensors is in prototype stage and will change in the near future. Please open a Github issue for features requests and see our documentation on the torch.masked module for further information about the project.
  return MaskedTensor(data, mask)
/home/zong/code/pytorch/torch/masked/maskedtensor/_ops_refs.py:354: UserWarning: The PyTorch API of MaskedTensors is in prototype stage and will change in the near future. Please open a Github issue for features requests and see our documentation on the torch.masked module for further information about the project.
  return MaskedTensor(new_data, _maybe_get_mask(args[0]))
Out[1]: (device(type='cuda', index=0), device(type='cuda', index=0))

In [2]: mt.sum(dim=0)
/home/zong/code/pytorch/torch/masked/maskedtensor/core.py:247: UserWarning: The PyTorch API of MaskedTensors is in prototype stage and will change in the near future. Please open a Github issue for features requests and see our documentation on the torch.masked module for further information about the project.
  return MaskedTensor(data, mask)
Out[2]: MaskedTensor(4, True)

```

```bash
pytest test/test_maskedtensor.py -vv
```

image

### CODE REVIEW

File: torch/masked/maskedtensor/core.py [hunk 1]
```diff
@@ -357,3 +357,17 @@ def is_sparse_csr(self):  # type: ignore[override]
     @property
     def is_sparse(self):
         return self.is_sparse_coo() or self.is_sparse_csr()
+
+    def to(self, *args, **kwargs):
+        current_device = self._masked_data.device
+        device_to = current_device
+        if len(args) == 1:
+            arg = args[0]
+            if isinstance(arg, (torch.device, str, int)):
```
Comment (Skylion007) [inline torch/masked/maskedtensor/core.py:366]: Use DeviceLIke from torch._prims_common and maybe use cannonicalize_device below?

Comment (zeshengzong) [reply torch/masked/maskedtensor/core.py]: Hi, do you mean using `DeviceLIkeType` like this?

```python
    def to(self, *args, **kwargs):
        current_device = self._masked_data.device
        device_to = current_device
        if len(args) == 1:
            arg = args[0]
            if isinstance(arg, DeviceLikeType):
                device_to = canonicalize_device(arg)
```

But mypy seems unhapply about it

```bash
>>> Lint for torch/masked/maskedtensor/core.py:

  Error (MYPY) [misc]
    Parameterized generics cannot be used with class or instance checks

        364  |        device_to = current_device
        365  |        if len(args) == 1:
        366  |            arg = args[0]
    >>> 367  |            if isinstance(arg, DeviceLikeType):
        368  |                device_to = canonicalize_device(arg)
        369  |            elif isinstance(arg, torch.Tensor):
        370  |                device_to = arg.device

  Error (MYPY) [arg-type]
    Argument 2 to "isinstance" has incompatible type "<typing special form>";
    expected "_ClassInfo"

        364  |        device_to = current_device
        365  |        if len(args) == 1:
        366  |            arg = args[0]
    >>> 367  |            if isinstance(arg, DeviceLikeType):
        368  |                device_to = canonicalize_device(arg)
        369  |            elif isinstance(arg, torch.Tensor):
        370  |                device_to = arg.device

```

And found a solution in here https://github.com/python/mypy/issues/12155#issuecomment-1660655381

**Comment by Skylion007** [comment]
Also, curious if it's easy to add tests in our existing test framework for this

File: torch/masked/maskedtensor/core.py [hunk 2]
```diff
@@ -357,3 +358,17 @@ def is_sparse_csr(self):  # type: ignore[override]
     @property
     def is_sparse(self):
         return self.is_sparse_coo() or self.is_sparse_csr()
+
+    def to(self, *args, **kwargs):
+        current_device = self._masked_data.device
+        device_to = current_device
+        if len(args) == 1:
+            arg = args[0]
+            if isinstance(arg, get_args(DeviceLikeType)):
```
[the remark this answers is not in the document]
Comment (zeshengzong) [reply torch/masked/maskedtensor/core.py]: Reverted, thanks!

**Comment by zeshengzong** [comment]
Hello @Skylion007 @cpuhrsch, is there anything else need to be changed before merge this PR, thanks!

**Comment by zeshengzong** [comment]
Hi @ezyang , please help review the change, thanks!

File: torch/masked/maskedtensor/core.py [hunk 3]
```diff
@@ -357,3 +358,17 @@ def is_sparse_csr(self):  # type: ignore[override]
     @property
     def is_sparse(self):
         return self.is_sparse_coo() or self.is_sparse_csr()
+
+    def to(self, *args, **kwargs):
```
Comment (ezyang) [inline torch/masked/maskedtensor/core.py]: I think this will work but there is something a bit suspicious.....

**Review by ezyang — CHANGES_REQUESTED** [review]
No, this seems wrong. Tensor.to semantics is out-of-place, we should not be in-place modifying the original tensor.

**Comment by ezyang** [comment]
It is still confusing to me why you have to override to. Why does the internal torch dispatch implementation not know how to transfer BOTH tensors. That is what the subclass's job is.

**Comment by fffrog** [comment]
Thanks @ezyang for the valuable suggestions.

The data and mask should be processed together by `__torch_dispatch__`, the root cause may be here, @zeshengzong can dig deeper and find out the reason.

https://github.com/pytorch/pytorch/blob/24b49b98810bb77f3cfa4c15baa9a15c9be3db61/torch/masked/maskedtensor/_ops_refs.py#L351-L354

**Comment by zeshengzong** [comment]
> Thanks @ezyang for the valuable suggestions.
> 
> The data and mask should be processed together by `__torch_dispatch__`, the root cause may be here, @zeshengzong can dig deeper and find out the reason.
> 
> https://github.com/pytorch/pytorch/blob/24b49b98810bb77f3cfa4c15baa9a15c9be3db61/torch/masked/maskedtensor/_ops_refs.py#L351-L354

Thanks! Made some changes in here and removed the override `to` method.

**Review by ezyang — APPROVED** [review]
